atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/CachingTransaction.java

```diff
@@ -110,57 +122,49 @@ public Transaction delegate() {
         }
     }
 
-    private void cacheLoadedRows(ConcurrentMap<Cell, byte[]> colCache,
-                                 Iterable<RowResult<byte[]>> rowView) {
+    private void cacheLoadedRows(Cache<Cell, byte[]> colCache,
+            Iterable<RowResult<byte[]>> rowView) {
         for (RowResult<byte[]> loadedRow : rowView) {
             for (Map.Entry<Cell, byte[]> e : loadedRow.getCells()) {
                 cacheLoadedColumns(colCache, ImmutableSet.of(e.getKey()), ImmutableMap.of(e.getKey(), e.getValue()));
             }
         }
     }
 
-    private void cacheLoadedColumns(ConcurrentMap<Cell, byte[]> colCache,
-                                    Set<Cell> toLoad,
-                                    Map<Cell, byte[]> toCache) {
+    private void cacheLoadedColumns(Cache<Cell, byte[]> colCache,
+            Set<Cell> toLoad,
+            Map<Cell, byte[]> toCache) {
         for (Cell key : toLoad) {
             byte[] value = toCache.get(key);
             if (value == null) {
                 value = PtBytes.EMPTY_BYTE_ARRAY;
             }
-            colCache.putIfAbsent(key, value);
+            colCache.put(key, value);
         }
     }
 
+    // Cache semantics: Zero byte values represent a deleted cell, which the delegate will return to us as null, but
+    // we will put in the cache as a zero-length array to differentiate "we saw a deleted cell" from "this cell isn't
+    // in the cache".
     @Override
     public Map<Cell, byte[]> get(TableReference tableRef, Set<Cell> cells) {
         if (cells.isEmpty()) {
             log.info("Attempted get on '{}' table with empty cells argument", tableRef);
             return ImmutableMap.of();
         }
 
-        ConcurrentMap<Cell, byte[]> cache = getColCacheForTable(tableRef);
-        Set<Cell> toLoad = Sets.newHashSet();
-        Map<Cell, byte[]> cacheHit = Maps.newHashMapWithExpectedSize(cells.size());
-        for (Cell cell : cells) {
-            byte[] val = cache.get(cell);
-            if (val != null) {
-                if (val.length > 0) {
-                    cacheHit.put(cell, val);
-                }
-            } else {
-                toLoad.add(cell);
-            }
+        try {
+            return getColCacheForTable(tableRef).getAll(cells).entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue().length > 0)
+                    .collect(Collectors.toConcurrentMap(entry -> entry.getKey(), entry -> entry.getValue()));
```

Comment:
No specific reason. I can make it a normal hash map if you think that's more reasonable.